File: api-node/src/healthcheck.js
/**
 * Sonda de salud del contenedor.
 *
 * Consulta GET /health contra el propio proceso y termina con codigo 0 si
 * responde correctamente, o 1 en cualquier otro caso.
 *
 * Existe por la misma razon que su equivalente en la API de Go: la imagen final
 * es distroless y no contiene shell, ni curl, ni wget con los que sondear.
 * Anadir alguna de esas herramientas solo para el healthcheck anularia el
 * motivo de usar distroless, que es no ofrecer utilidades a quien logre entrar
 * en el contenedor.
 *
 * Uso:  node src/healthcheck.js
 */

const port = process.env.PORT || '3001';

// La sonda debe fallar rapido: si el servicio no responde en dos segundos,
// esperar mas solo retrasa la deteccion de que algo va mal.
const TIMEOUT_MS = 2000;

// Se consulta 127.0.0.1 y no el nombre del servicio: la sonda comprueba ESTE
// proceso, no la resolucion de nombres de la red del contenedor.
const url = `http://127.0.0.1:${port}/health`;

try {
  const response = await fetch(url, { signal: AbortSignal.timeout(TIMEOUT_MS) });

  if (!response.ok) {
    process.stderr.write(`healthcheck: estado ${response.status}\n`);
    process.exit(1);
  }

  process.exit(0);
} catch (error) {
  process.stderr.write(`healthcheck: sin respuesta: ${error.message}\n`);
  process.exit(1);
}

File: api-go/cmd/api/healthcheck.go
package main

import (
	"context"
	"fmt"
	"net/http"
	"os"
	"time"
)

// healthcheckCommand es el nombre del subcomando de sonda.
const healthcheckCommand = "healthcheck"

// runHealthcheck consulta el endpoint de salud del propio proceso y termina
// con codigo 0 si responde correctamente, o 1 en cualquier otro caso.
//
// Existe porque la imagen final es distroless: no contiene shell, ni curl, ni
// wget con los que sondear el servicio. Anadir alguna de esas herramientas
// solo para el healthcheck echaria por tierra la razon de usar distroless, que
// es precisamente no ofrecer utilidades a quien logre entrar en el contenedor.
//
// La alternativa idiomatica es la que se implementa aqui: el mismo binario que
// sirve la API sabe tambien comprobarla. No anade superficie de ataque —el
// ejecutable ya esta en la imagen— y funciona igual en Docker, en Compose y en
// Kubernetes.
//
// Uso:  /app/api healthcheck
func runHealthcheck() int {
	port := os.Getenv("PORT")
	if port == "" {
		port = "8080"
	}

	// La sonda debe fallar rapido: si el servicio no responde en dos segundos,
	// esperar mas solo retrasa la deteccion de que algo va mal.
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()

	// Se consulta 127.0.0.1 y no el nombre del servicio: la sonda comprueba
	// ESTE proceso, no la resolucion de nombres de la red.
	url := fmt.Sprintf("http://127.0.0.1:%s/health", port)

	request, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		fmt.Fprintf(os.Stderr, "healthcheck: peticion invalida: %v\n", err)
		return 1
	}

	response, err := http.DefaultClient.Do(request)
	if err != nil {
		fmt.Fprintf(os.Stderr, "healthcheck: sin respuesta: %v\n", err)
		return 1
	}
	defer response.Body.Close()

	if response.StatusCode != http.StatusOK {
		fmt.Fprintf(os.Stderr, "healthcheck: estado %d\n", response.StatusCode)
		return 1
	}

	return 0
}
